src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SecurityDirectives.scala

```diff
@@ -18,30 +19,37 @@ trait SecurityDirectives {
-        else
-          Directives.failWith(
-            new WorkbenchExceptionWithErrorReport(ErrorReport(
-              StatusCodes.Forbidden,
-              s"You may not perform any of ${requestedActions.mkString("[", ", ", "]").toString.toUpperCase} on ${resource.resourceTypeName.value}/${resource.resourceId.value}"
-            )))
+      onSuccess(hasPermissionOneOf(resource, requestedActions, userId)) { hasPermission =>
+        if (hasPermission) {
+          innerRoute
+        } else {
+
+          // in the case where we don't have the required action, we need to figure out if we should return
+          // a Not Found (you have no access) vs a Forbidden (you have access, just not the right kind)
+          onSuccess(policyEvaluatorService.listResourceAccessPoliciesForUser(resource, userId)) { policies =>
+
+            if (policies.isEmpty) {
+              Directives.failWith(
+                new WorkbenchExceptionWithErrorReport(
+                  ErrorReport(StatusCodes.NotFound, s"Resource ${resource.resourceTypeName.value}/${resource.resourceId.value} not found")))
+            } else {
+              Directives.failWith(
+                new WorkbenchExceptionWithErrorReport(ErrorReport(
+                  StatusCodes.Forbidden,
+                  s"You may not perform any of ${requestedActions.mkString("[", ", ", "]").toString.toUpperCase} on ${resource.resourceTypeName.value}/${resource.resourceId.value}"
+                )))
+            }
+          }
+        }
       }
     }
 
-  private def listActions(resource: FullyQualifiedResourceId, userId: WorkbenchUserId, requestedActions: Set[ResourceAction]): IO[Set[ResourceAction]] =
-    // this is optimized for the case where the user has permission since that is the usual case
-    // if the first attempt shows the user does not have permission, force a second attempt
+  private def hasPermissionOneOf(resource: FullyQualifiedResourceId, actions: Iterable[ResourceAction], userId: WorkbenchUserId): IO[Boolean] =
+    //  first quickly check if we have permission using the shallow check across all actions, then try the full check
     for {
-      actionsAttempt1 <- policyEvaluatorService.listUserResourceActions(resource, userId, force = false)
-      actionsAttempt2 <- if (hasAccess(requestedActions, actionsAttempt1)) IO.pure(actionsAttempt1)
-      else policyEvaluatorService.listUserResourceActions(resource, userId, force = true)
-    } yield actionsAttempt2
-
-  private def hasAccess(requestedActions: Set[ResourceAction], actions: Set[ResourceAction]): Boolean =
-    actions.intersect(requestedActions).nonEmpty
+      attempt1 <- actions.toList.existsM(policyEvaluatorService.hasPermissionShallowCheck(resource, _, userId))
+      attempt2 <- if (attempt1) IO.pure(attempt1) else actions.toList.existsM(policyEvaluatorService.hasPermissionFullCheck(resource, _, userId))
```

Comment:
Agreed about the regression. In the case where the user does not have access this will do more ismemberof calls than before. In reality this looks like there are at most 2 actions ever required and almost all of those are in the policy altering apis. So this may be an acceptable regression.

In the case where the user does have permission but not shallowly assigned, caching should ensure there is only 1 ismemberof call.